src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs

```diff
@@ -2012,28 +2019,26 @@ private static string CheckForMatches(uint[] runningHash, int upTo)
             /// <returns>The string matching the hash, or null.</returns>
             public static string Match(string text)
```

Comment:
When reviewing the changes, I found the algorithm in `Match` has flaws: when the text starts with a suspicious text defined in `LookupHash(uint h)`, it cannot find the text.

#### dotnet application to reproduce the issue

```c#
      static void Main(string[] args)
      {
          string match = SuspiciousContentChecker.Match("Add-Type");
          if (match is null)
          {
              Console.WriteLine("match not found.");
          }
          else
          {
              Console.WriteLine($"match found: {match}");
          }
      }

     // Copy the SuspiciousContentChecker class here.
     // ....
```

Results from running the application:

![image](https://user-images.githubusercontent.com/127450/204669157-fed84338-b03a-4ff6-9e4b-7335a4ef012b.png)

To make the `Match` work, at least 1 more character needs to be present before `Add-Type`. Change it to `SuspiciousContentChecker.Match(" Add-Type");`, then it's able to find the `Add-Type` text:

![image](https://user-images.githubusercontent.com/127450/204669585-324c0636-f0f7-40ee-9915-6c129369cf8e.png)


----

The `Match` call is used in [CheckSuspiciousContent(Ast scriptBlockAst)](https://github.com/PowerShell/PowerShell/blob/d03a5553c237c39dcbe7fac0c1d39eef9f140ec6/src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs#LL1775-L1779C14):
```
            var foundSignature = SuspiciousContentChecker.Match(scriptBlockAst.Extent.Text);
            if (foundSignature != null)
            {
                return foundSignature;
            }
```
And of course, we can create a script block starts with `Add-Type`, for example: 

![image](https://user-images.githubusercontent.com/127450/204670373-412f10f8-8312-489d-91a8-a0b434a2fb74.png)

Reply:
I think the fix is: instead of
```
for (int j = Math.Min(i, runningHash.Length) - 1; j > 0; j--)
```
we should remove the `- 1` part:
```
for (int j = Math.Min(i, runningHash.Length); j > 0; j--)
```